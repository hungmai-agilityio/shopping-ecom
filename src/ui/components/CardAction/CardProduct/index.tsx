'use client';

import { ICart, IUser, IWishlist } from '@/interface';
import { memo, useCallback, useState } from 'react';
import { v4 as uuidv4 } from 'uuid';
import { useRouter } from 'next/navigation';

// Hooks
import {
  useAddDataToCart,
  useAddToWishlist,
  useRemoveFromWishlist,
  useUpdateDataToCart
} from '@/hooks';

// Components
import { CardProduct, ToastMessage } from '@/ui/components';

// Libs
import { getUserCart, getUserWishList } from '@/libs';

// Constants
import { END_POINT, MESSAGE_API, QUERY, STATUS } from '@/constants';
import { useQuery } from '@tanstack/react-query';

interface CardProductActionProps {
  colors: string[];
  discount?: number;
  id: string;
  image: string;
  isDiscount?: boolean;
  isNew?: boolean;
  isNewProduct?: boolean;
  name: string;
  originalPrice?: number;
  price: number;
  ratings: number;
  user: IUser;
}

const CardProductAction = memo(
  ({
    colors = [],
    discount,
    id,
    image,
    isDiscount,
    isNew,
    isNewProduct,
    name,
    originalPrice,
    price,
    ratings,
    user
  }: CardProductActionProps) => {
    const [toast, setToast] = useState<{
      status: STATUS;
      message: string;
    } | null>(null);

    const addToWishlist = useAddToWishlist();
    const removeFromWishlist = useRemoveFromWishlist();
    const updateDataToCart = useUpdateDataToCart();
    const addToCart = useAddDataToCart();
    const router = useRouter();

    const { data: wishlist = [] } = useQuery<IWishlist[]>({
      queryKey: [QUERY.WISHLIST],
      queryFn: () => getUserWishList(user.id),
      enabled: !!user
    });

    const isProductInWishlist = (productId: string) =>
      wishlist.some((item) => item.productId === productId);

    // Handle toggle add/remove product favorite
    const handleToggleFavorite = useCallback(() => {
      if (isProductInWishlist(id)) {
        const wishlistItem = wishlist.find((item) => item.productId === id);
        if (wishlistItem) {
          removeFromWishlist.mutate(wishlistItem.id);
        }
        return;
      }

      const newItem: IWishlist = {
        id: uuidv4(),
        userId: user.id,
        productId: id
      };

      addToWishlist.mutate(newItem);
    }, [id, user, wishlist, removeFromWishlist, addToWishlist]);

    /**
     * Handle adding a product to the cart.
     * If the product already exists in the cart, its quantity is incremented by 1.
     * If not, a new cart item is added.
     * @param {string} productId
     */
    const handleAddToCart = useCallback(
      async (productId: string) => {
        const cartItems = await getUserCart(user!.id);

        const existingItem = cartItems.find(
          (cartItem: ICart) =>
            cartItem.productId === productId && cartItem.userId === user?.id
        );

        if (existingItem) {
          updateDataToCart.mutate({
            id: existingItem.id,
            data: { quantity: existingItem.quantity + 1 }
          });

          setToast({
            status: STATUS.SUCCESS,
            message: MESSAGE_API.UPDATE_QUANTITY
          });
        }
        if (!existingItem) {
          const newItem: ICart = {
            id: uuidv4(),
            userId: user.id,
            productId: productId,
            color: '',
            size: '',
            quantity: 1
          };
          addToCart.mutate(newItem);
          setToast({
            status: STATUS.SUCCESS,
            message: MESSAGE_API.ADD_PRODUCT_SUCCESS
          });
        }
      },
      [user]
    );

    const handleRedirectPreview = (id: string) => {
      router.push(`${END_POINT.PRODUCT}/${id}`);
    };

    return (
      <>
        <CardProduct
          alt={name}
          colors={colors}
          discount={isDiscount ? discount : undefined}
          icon={isProductInWishlist(id) ? '/heart-red.svg' : '/heart.svg'}
          id={id}
          isNewProduct={isNewProduct ? isNew : undefined}
          isShowAction={!!user}
          key={id}
          name={name}
          oldPrice={originalPrice}
          onAdd={handleAddToCart}
          onIconClick={handleToggleFavorite}
          price={price}
          rating={ratings}
          selectedColor={colors[0]}
          src={image}
          onView={handleRedirectPreview}
        />
        {toast && (
          <ToastMessage status={toast.status} message={toast.message} />
        )}
      </>
    );
  }
);

export default CardProductAction;
