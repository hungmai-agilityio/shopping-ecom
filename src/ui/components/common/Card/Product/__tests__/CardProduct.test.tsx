describe('CardProduct component', () => {});

import CardProduct from '@/ui/components/common/Card/Product';
import { render, screen, fireEvent } from '@testing-library/react';

describe('CardProduct', () => {
  const defaultProps = {
    src: 'https://i.ibb.co/Jx5zGXk/head-1.webp',
    alt: 'Example Product Image',
    id: 'product-1',
    name: 'Example Product',
    colors: ['Red', 'Blue'],
    rating: 4,
    price: 120,
    oldPrice: 150,
    icon: '/heart.svg',
    selectedColor: 'Red',
    onAdd: jest.fn(),
    onIconClick: jest.fn(),
    onView: jest.fn(),
    isShowAction: true
  };

  test('Should render with default props', () => {
    render(<CardProduct {...defaultProps} />);

    expect(screen.getByText('Example Product')).toBeInTheDocument();
    expect(screen.getByText('$120')).toBeInTheDocument();
    expect(screen.getByText('$150')).toBeInTheDocument();
    expect(screen.getByText('(4)')).toBeInTheDocument();
  });

  test('Should displays discount label when discount props', () => {
    render(<CardProduct {...defaultProps} discount={20} />);

    expect(screen.getByText('-20%')).toBeInTheDocument();
  });

  test('displays NEW label when isNewProduct is true', () => {
    render(<CardProduct {...defaultProps} isNewProduct />);

    expect(screen.getByText('NEW')).toBeInTheDocument();
  });

  test('Test function onAdd when Add To Cart button is clicked', () => {
    render(<CardProduct {...defaultProps} />);
    fireEvent.mouseEnter(screen.getByAltText('Example Product Image'));
    const addToCartButton = screen.getByText('Add To Cart');
    expect(addToCartButton).toBeInTheDocument();
    fireEvent.click(addToCartButton);

    expect(defaultProps.onAdd).toHaveBeenCalledTimes(1);
  });

  test('Test function onIconClick when icon is clicked', () => {
    render(<CardProduct {...defaultProps} />);

    fireEvent.click(screen.getByAltText('icon-click'));

    expect(defaultProps.onIconClick).toHaveBeenCalledTimes(1);
  });

  test('Test function onView function when View icon is clicked', () => {
    render(<CardProduct {...defaultProps} />);

    const viewIcon = screen.getByAltText('Example Product Image');
    fireEvent.click(viewIcon);

    expect(defaultProps.onView).toHaveBeenCalledTimes(1);
  });

  test('Should renders color picker with available colors', () => {
    render(<CardProduct {...defaultProps} />);

    expect(screen.getByLabelText('Red-picker')).toBeInTheDocument();
    expect(screen.getByLabelText('Blue-picker')).toBeInTheDocument();
  });

  test('Should render match to snapshot', () => {
    const { container } = render(<CardProduct {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });
});
