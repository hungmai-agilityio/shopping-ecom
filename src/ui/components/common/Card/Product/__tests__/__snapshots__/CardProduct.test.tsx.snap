// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardProduct Should render match to snapshot 1`] = `
<div>
  <div
    class="rounded-sm overflow-hidden className w-card-sm h-card-product font-medium"
  >
    <div
      class="bg-gray-light p-4 rounded-b-sm h-card-md relative flex group"
    >
      <div
        class="w-9 h-9 absolute top-3 right-5"
      >
        <div
          class="flex bg-white p-2 rounded-full cursor-pointer"
        >
          <img
            alt="icon-click"
            data-nimg="1"
            decoding="async"
            fetchpriority="high"
            height="34"
            src="/heart.svg"
            style="color: transparent;"
            width="34"
          />
        </div>
      </div>
      <div
        class="w-9 h-9 absolute right-5 top-14"
      >
        <div
          class="flex bg-white p-2 rounded-full cursor-pointer"
        >
          <img
            alt="icon-view"
            data-nimg="1"
            decoding="async"
            fetchpriority="high"
            height="34"
            src="/eye.svg"
            style="color: transparent;"
            width="34"
          />
        </div>
      </div>
      <img
        alt="Example Product Image"
        class="mx-auto"
        data-nimg="1"
        decoding="async"
        height="100"
        loading="lazy"
        src="/_next/image?url=https%3A%2F%2Fi.ibb.co%2FJx5zGXk%2Fhead-1.webp&w=384&q=75"
        srcset="/_next/image?url=https%3A%2F%2Fi.ibb.co%2FJx5zGXk%2Fhead-1.webp&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fi.ibb.co%2FJx5zGXk%2Fhead-1.webp&w=384&q=75 2x"
        style="color: transparent; object-fit: contain;"
        width="152"
      />
      <div
        class="absolute bg-dark w-full h-10 group-hover:flex justify-center items-center gap-4 hover:bg-slate-600 hidden bottom-0 left-0 cursor-pointer"
      >
        <div
          class="flex"
        >
          <img
            alt="icon-cart"
            data-nimg="1"
            decoding="async"
            fetchpriority="high"
            height="24"
            src="/cart-white.svg"
            style="color: transparent;"
            width="24"
          />
        </div>
        <p
          class="text-white"
        >
          Add To Cart
        </p>
      </div>
    </div>
    <div
      class="mt-4 px-2"
    >
      <p
        class="text-base overflow-hidden line-clamp-1 capitalize"
      >
        Example Product
      </p>
      <div
        class="mt-2 flex items-center gap-3"
      >
        <span
          class="text-primary max-w-16 overflow-hidden text-ellipsis inline-block"
        >
          $
          120
        </span>
        <span
          class="text-gray-500 line-through max-w-16 overflow-hidden text-ellipsis inline-block"
        >
          $
          150
        </span>
        <div
          class="flex gap-1"
        >
          <img
            alt="rating 1"
            data-nimg="1"
            decoding="async"
            fetchpriority="high"
            height="20"
            src="/star.svg"
            style="color: transparent;"
            width="20"
          />
          <img
            alt="rating 2"
            data-nimg="1"
            decoding="async"
            fetchpriority="high"
            height="20"
            src="/star.svg"
            style="color: transparent;"
            width="20"
          />
          <img
            alt="rating 3"
            data-nimg="1"
            decoding="async"
            fetchpriority="high"
            height="20"
            src="/star.svg"
            style="color: transparent;"
            width="20"
          />
          <img
            alt="rating 4"
            data-nimg="1"
            decoding="async"
            fetchpriority="high"
            height="20"
            src="/star.svg"
            style="color: transparent;"
            width="20"
          />
        </div>
        <span
          class="text-gray-500"
        >
          (
          4
          )
        </span>
      </div>
    </div>
    <div
      class="mt-2 px-2"
    >
      <div
        class="flex items-center gap-3"
      >
        <button
          aria-label="Red-picker"
          class="h-4 w-4 rounded-full focus:border focus:border-green-800 ring-2 ring-dark"
          id="Red"
          style="background-color: Red;"
        />
        <button
          aria-label="Blue-picker"
          class="h-4 w-4 rounded-full focus:border focus:border-green-800"
          id="Blue"
          style="background-color: Blue;"
        />
      </div>
    </div>
  </div>
</div>
`;
